_posts/2022-00-00-signature-polymorphic-methods.md

```diff
@@ -0,0 +1,249 @@
+some future JVM version?)
+
+## What does "signature polymorphism" mean, exactly?
+
+As the [Javadoc for `MethodHandle`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/MethodHandle.html) says,
+
+> A signature polymorphic method is one which can operate with any of
+> a wide range of call signatures and return types.
+
+and:
+
+> In source code, a call to a signature polymorphic method will
+> compile, regardless of the requested symbolic type descriptor. As
+> usual, the Java compiler emits an invokevirtual instruction with the
+> given symbolic type descriptor against the named method. The unusual
+> part is that the symbolic type descriptor is derived from the actual
+> argument and return types, not from the method declaration.
+
+Note that generics are not sufficient to express this level of
+flexibility, for two reasons:
+
+First, Java generics only work on reference types, not primitive
+types.  Scala does not have this limitation, but pays for it by
+incurring boxing at run-time when primitive types are used in generic
+contexts.
+
+Second, methods (in both languages) may only have a fixed number of
+type parameters.
+
+The following example should help make all of this clearer.
+
+## How do I call a signature polymorphic method from Scala?
+
+Take `MethodHandle` for example. It provides an `invokeExact`
+method. Its signature as seen from Scala is:
+
+    def invokeExact(args: AnyRef*): AnyRef
+
+Signature polymorphism means that the `AnyRef`s here are just
+placeholders for types to be supplied later.
+
+To see this work in practice, let's adapt one of the examples in
+the Javadoc. From Scala, we'll make a reflective call to the `replace`
+method on a `String`:
+
+    import java.lang.invoke._
+    val mt = MethodType.methodType(
+      classOf[String], classOf[Char], classOf[Char])
+    val mh = MethodHandles.lookup.findVirtual(
+      classOf[String], "replace", mt)
+    val s = mh.invokeExact("daddy", 'd', 'n'): String
+
+If we paste this into the Scala REPL (2 or 3), we see:
+
+    val s: String = nanny
+
+Signature polymorphism helped us here in two ways:
+
+* The arguments `d` and `n` will not be passed as `Object` or boxed to
+  `java.lang.Character` at runtime, but will be passed directly as
+  primitive `Char`s.
+* The result comes back as a `String` without needing to be checked
+  or cast at runtime.
+
+## Are these methods good for anything else?
+
+Great question!
+
+Doesn't it seem puzzling that Oracle would go to so much trouble to
+make Java reflection faster? If I care so much about performance,
+surely I should avoid using reflection at all?
+
+The real reason these methods need to be fast is to aid efficient
+implementation of lambdas, in both Java and Scala. Additionally, they aid
+efficient implementation of dynamic languages on the
+JVM. `MethodHandle` was added to the JVM at the same time as
+`invokeDynamic`, which serves those same two purposes.
```

Comment:
Maybe you can link to [JSR-292](https://jcp.org/en/jsr/detail?id=292) proposal?

> To make it easier to produce performant, high quality implementations of [dynamically typed] languages, we propose to add support at the virtual machine level.
>
> Specifically, we seek to add a new JVM instruction, `invokedynamic`, designed to support the implementation of dynamically typed object oriented languages.

Reply:
See also https://wiki.openjdk.org/display/mlvm/InterfaceDynamic

Reply:
I decided to link to https://www.infoq.com/articles/invokedynamic/ , which seems like a good introduction and links to further information.